src/pineko/kfactor.py

```diff
@@ -209,148 +193,108 @@ def create_singlegridonly(
-        grid_path,
-        max_as,
-        first_nonzero_order[0],
-        min_al,
-        centrals_kfactor,
-        alphas,
-        order_exists,
-    )
-    # Writing the sv grids
-    grids_paths = scale_variations.write_grids(gridpath=grid_path, grid_list=grid_list)
-    # Merging all together
-    scale_variations.merge_grids(
-        gridpath=grid_path,
-        grid_list_path=grids_paths,
-        target_path=target_folder,
-        nec_orders=nec_orders,
-        order_exists=order_exists,
-    )
+    # the actual alpha_s order to update
+    order_to_update = pto_to_update + min_as - 1
 
+    # check if the order is already there
+    is_in = is_already_in((order_to_update, min_al, 0, 0), grid_orders_filtered)
 
-def do_it(
-    centrals_kfactor,
-    alphas,
-    grid_path,
-    grid,
-    max_as,
-    max_as_test,
-    target_folder,
-    order_exists,
-):
-    """Apply the centrals_kfactor to the grid if the order is not already there."""
-    grid_orders = [orde.as_tuple() for orde in grid.orders()]
-    order_mask = pineappl.grid.Order.create_mask(grid.orders(), max_as, 0, True)
-    grid_orders_filtered = list(np.array(grid_orders)[order_mask])
-    grid_orders_filtered.sort(key=scale_variations.qcd)
-    first_nonzero_order = grid_orders_filtered[0]
-    min_al = first_nonzero_order[1]
-    is_in = is_already_in(
-        (first_nonzero_order[0] + max_as_test, min_al, 0, 0), grid_orders_filtered
-    )
+    # Prevent summing orders incoherently
     if is_in and not order_exists:
         rich.print("[green] Success: Requested order already in the grid.")
         return
     if not is_in and order_exists:
         rich.print("[red] Abort: order exists is True but order not in the grid.")
         return
-    construct_and_merge_grids(
-        grid_path,
-        max_as_test,
-        first_nonzero_order,
-        min_al,
-        centrals_kfactor,
-        alphas,
-        target_folder,
-        order_exists,
-    )
+
+    # loop on all the order to update
+    max_as = grid_orders_filtered[-1][0]
+    orders_list = [(de, min_al, 0, 0) for de in range(min_as, max_as + 1)]
+    # create an empty grid and add the rescaled order
+    order_to_update = (order_to_update, min_al, 0, 0)
+    new_order_grid = None
+    for i, as_order in enumerate(orders_list):
+        order_grid = construct_new_order(
+            grid, as_order, order_to_update, central_kfactor, alphas
+        )
+        if i == 0:
+            new_order_grid = order_grid
+        else:
+            new_order_grid.merge(order_grid)
+
+    new_grid = grid
+    # if the new order is there, clean the old one.
+    if is_in:
+        new_grid = scale_variations.construct_and_dump_order_exists_grid(
+            grid, order_to_update
+        )
+    # merge the updated order with the original one.
+    new_grid.merge(new_order_grid)
+    new_grid.write_lz4(target_grid_path)
 
 
 def filter_k_factors(pigrid, centrals_kfactor):
```

Comment:
```suggestion
def filter_k_factors(grid, central_kfactors):
```
? (+ all the implications ...)

Reply:
done in [3163be4](https://github.com/NNPDF/pineko/pull/161/commits/3163be4955b25bc48118002614d3f799284fa3f1)

Reply:
see [55e015d](https://github.com/NNPDF/pineko/pull/161/commits/55e015d7f60753c0982d059cbfc6fb4c32c37aab)